File: src/contexts/core/kanji/domain/entities/kanji.ts
import { deepEquals } from "bun";
import type { DomainEvent } from "../../../../shared/events/domain/domain-event";
import { DomainEventId } from "../../../../shared/events/domain/domain-event-id";
import { DomainEventOccurredOn } from "../../../../shared/events/domain/domain-event-occurred-on";
import { KanjiCreatedEvent } from "../events/kanji-created-event";
import type { KanjiIdeogram } from "../valueobjects/kanji-ideogram";
import type { KanjiKunyomiReading } from "../valueobjects/kanji-kunyomi-reading";
import type { KanjiMeaning } from "../valueobjects/kanji-meaning";
import type { KanjiOnyomiReading } from "../valueobjects/kanji-onyomi-reading";
import type { KanjiRadical } from "../valueobjects/kanji-radical";
import type { KanjiStrokesNumber } from "../valueobjects/kanji-strokes-number";
import { AggregateRoot } from "../../../../shared/aggregate/domain/aggregate-root";

class Kanji extends AggregateRoot {
    constructor(
        public ideogram: KanjiIdeogram,
        public onyomiReadings: KanjiOnyomiReading[],
        public kunyomiReadings: KanjiKunyomiReading[],
        public meanings: KanjiMeaning[],
        public strokes: KanjiStrokesNumber,
        public radicals: KanjiRadical[],
    ) {
        super();
    }

    static create(
        ideogram: KanjiIdeogram,
        onyomiReadings: KanjiOnyomiReading[],
        kunyomiReadings: KanjiKunyomiReading[],
        meanings: KanjiMeaning[],
        strokes: KanjiStrokesNumber,
        radicals: KanjiRadical[],
    ): [Kanji, DomainEvent] {
        const kanji = new Kanji(ideogram, onyomiReadings, kunyomiReadings, meanings, strokes, radicals);
        const event = new KanjiCreatedEvent(DomainEventId.random(), DomainEventOccurredOn.now(), ideogram, onyomiReadings, 
            kunyomiReadings, meanings, strokes, radicals);
        return [kanji, event];
    }

    toPrimitives(): any {
        const data = {
            ideogram: this.ideogram.val,
            onyomiReadings: this.onyomiReadings.map((reading) => reading.val),
            kunyomiReadings: this.kunyomiReadings.map((reading) => reading.val),
            meanings: this.meanings.map((meaning) => meaning.val),
            strokes: this.strokes.val,
            radicals: this.radicals.map((radical) => radical.val),
        }
        return data;
    }

    equals(kanji: Kanji) {
        return deepEquals(this, kanji);
    }
}

export { Kanji };

File: libs/kernel/src/kanjis/domain/entities/kanji.rs
use aggregate_root::domain::aggregate_root::AggregateRoot;

use super::{kanji_ideogram::KanjiIdeogram, kanji_kunyomi_reading::KanjiKunyomiReading, kanji_meaning::KanjiMeaning, kanji_nanori_reading::KanjiNanoriReading, kanji_onyomi_reading::KanjiOnyomiReading, kanji_radical::KanjiRadical, kanji_strokes_number::KanjiStrokesNumber};


pub struct Kanji {
    pub ideogram: KanjiIdeogram,
    pub kunyomi_readings: Vec<KanjiKunyomiReading>,
    pub onyomi_readings: Vec<KanjiOnyomiReading>,
    pub nanori_readings: Vec<KanjiNanoriReading>,
    pub meanings: Vec<KanjiMeaning>,
    pub strokes: KanjiStrokesNumber,
    pub radicals: Vec<KanjiRadical>,
}

impl Kanji {
    pub fn new(ideogram: KanjiIdeogram, kunyomi_readings: Vec<KanjiKunyomiReading>, onyomi_readings: Vec<KanjiOnyomiReading>, 
        nanori_readings: Vec<KanjiNanoriReading>, meanings: Vec<KanjiMeaning>, strokes: KanjiStrokesNumber, radicals: Vec<KanjiRadical>) -> Self {
        Self { ideogram, kunyomi_readings, onyomi_readings, nanori_readings, meanings, strokes, radicals }
    }
}

impl AggregateRoot for Kanji {
    fn get_type() -> String {
        "tanukanji.kernel.kanji".to_string()
    }
}

impl Clone for Kanji {
    fn clone(&self) -> Self {
        Self::new(self.ideogram.clone(), self.kunyomi_readings.clone(), self.onyomi_readings.clone(), self.nanori_readings.clone(), self.meanings.clone(), self.strokes.clone(), self.radicals.clone())
    }
}
